librosa/sequence.py

```diff
@@ -208,7 +220,11 @@ def dtw(X=None, Y=None, C=None, metric='euclidean', step_sizes_sigma=None,
 
     # initialize step matrix with -1
     # will be filled in calc_accu_cost() with indices from step_sizes_sigma
-    D_steps = -1 * np.ones(D.shape, dtype=np.int)
+    D_steps = np.zeros(D.shape, dtype=np.int)
+    # TODO: replace these with steps that correspond to left (first row)
+    # and up (first column)
+    D_steps[0, :] = 1
+    D_steps[:, 0] = 2
```

Comment:
@fzalkow @stefan-balke Any ideas for how to address this one?

The old version (-1) initializes all steps to whatever the last entry in the `step_sizes` array is, with the assumption that every element of `D_steps` will be overwritten before any backtracking is attempted.  However, this does not always happen, eg when there are `Inf` values in the cost matrix.

The current change (not intended to be a permanent fix) sets the initial step values in the first row and column to be horizontal and vertical moves respectively, but this is brittle and will fail if alternate step sizes are provided.  I'd like to have a robust solution that works in general, but don't see how to achieve that without adding restrictions on the step sizes.  Any thoughts?

Reply:
Instead of 1 and 2, we can use `most_horizontal_step_idx` and `most_vertical_step_idx` (see my PR #1033). Still, it may cover not all cases with arbitrary `infs` in the cost matrix.

Reply:
I agree in general, but I don't think steepest/shallowest slope is a strong enough condition.  What we want to do here is prevent the back-pointers from going off the edge of the array, which means the first/row column has to be purely horizontal or vertical moves.

Alternately, we could rewrite the data structure such that `D_steps` contains the absolute position instead of the index into the step_sizes array.  This would allow us to populate the steps matrix with technically illegal moves that form a coherent path, and then overwrite them when better solutions are available.  Doing this would require a bit of numpy hackery though, and possibly a custom structure dtype for tuples, but it might be the cleanest solution in the long run.

Reply:
I changed my PR #1033 so that now we check the accumulated cost matrix before doing backtracking. If we additionally forbid nans in the cost matrix (infs are fine!), I can see no option for the backtracking to access anything outside the array edges.

Reply:
@fzalkow I gave this some thought, and I have an idea that could work.

Instead of replacing the default step sizes (and costs/weights) with user-supplied values, we can keep the defaults ([1,1], [0,1], [1,0]) and append user-supplied steps to that list.  If that happens, we set the weights for the defaults to infinity so that they are never preferred over the user-supplied steps.  It's a little hacky, but I think it nicely solves our problems with initializing the step matrix.

We'd still have the potential problem of running off the edge of the array while backtracking, so that special-case handler still needs to be there.

What do you think?

Reply:
Implemented this and added a safety check for negative step parameters.  Tests pass, and i'm reasonably happy with the solution.  Please give it a look over!

Reply:
I don't know if it is a good idea that non-supplied step sizes can be used. For example, in the following case, I would expect DTW not to turn any result (and in particular, not to use any non-supplied step sizes), because there is no valid solution given the supplied step sizes.

```python
step_sizes_sigma = np.array([[1, 1], [1, 2], [2, 1]])
C = np.ones((4, 2))
D, wp = librosa.sequence.dtw(C=C, step_sizes_sigma=step_sizes_sigma)
```

In the this current branch, full DTW returns no valid warping path for this example (not using cell `(0, 0)`), because of the comparison between costs of the added infinity rows/columns and the actual infinity values of the `D` matrix.

I would rather be in favor of raising an exception as in PR #1033.

Just my opinion, of course!

Reply:
> I don't know if it is a good idea that non-supplied step sizes can be used.

The whole point is that they won't be used -- any path with a finite cost will take precedent.  All this proposed change does is ensure that the step matrix is always well-formed.


>  would expect DTW not to turn any result (and in particular, not to use any non-supplied step sizes), because there is no valid solution given the supplied step sizes.

Agreed.  The current branch does not yet implement the error-handling logic, so the call succeeds.  I think I see what you're saying with the example though; there's no explicit logic to distinguish valid (finite-cost) paths from invalid (infinite-cost) paths.  I think we could easily add that to the loop condition in the backtracker and that should resolve everything.

Reply:
> The whole point is that they won't be used -- any path with a finite cost will take precedent. All this proposed change does is ensure that the step matrix is always well-formed.

I am not sure if we have full control over that. When initializing the `D` matrix, additional infinity columns and rows are created outside of the valid range of `D` ([code link](https://github.com/librosa/librosa/blob/90ef6b2736c127bfb7e314b8b839fd5b494f52ce/librosa/sequence.py#L243)). Now, if steps with infinite cost are introduced, then a comparison of infinity to infinity may be performed (going outside the valid range or using a non-supplied step).

Sorry for bringing it up again. :-) PR #1033 proposes not to do any backtacking when the upper right corner of the `D` matrix has infinite cost. Shouldn't that solve the overall issue?

Reply:
>  then a comparison of infinity to infinity may be performed (going outside the valid range or using a non-supplied step).

The step value is only updated if the proposed cost is less than the initial cost; `inf < inf` is False, so the initial value will remain in place, guaranteeing pointer safety.



> PR #1033 proposes not to do any backtracking when the upper right corner of the `D` matrix has infinite cost. Shouldn't that solve the overall issue?

I think it does in common cases, but I'm more worried at this point about correctly handling non-standard step sizes.  For example, what if a user wants to forbid diagonal moves and use `step_sizes_sigma = [[0, 1], [1, 0]]`?  In this case, initializing steps with `1, 2` would lead to segfaults again.  More generally, i'd really like to avoid anything that involves scanning through the step sizes matrix to find "good" initialization values.

Reply:
> > PR #1033 proposes not to do any backtracking when the upper right corner of the `D` matrix has infinite cost. Shouldn't that solve the overall issue?
> 
> I think it does in common cases, but I'm more worried at this point about correctly handling non-standard step sizes. For example, what if a user wants to forbid diagonal moves and use `step_sizes_sigma = [[0, 1], [1, 0]]`? In this case, initializing steps with `1, 2` would lead to segfaults again.

As far as I see, it would not lead to segfaults. If there is a valid solution for these step sizes (meaning `D[-1, -1]` is not infinity), then all used entries in `D_steps` are overwritten before backtracking. If there is no valid solution, then the backtracking won't be performed. Maybe I'm overlooking something here?

Reply:
Hrm, I suppose you're right.  It still makes me nervous having bad values in the step matrix though, even if there's logic elsewhere that (indirectly) prevents them from being used as pointer offsets.

I'm still thinking over how this relates to #1033 though -- I still don't think they're addressing the same issue exactly, but I'm having a difficult time articulating what exactly the difference is.

Reply:
Coming back to this after some time away, I think this discussion got confusing because we're trying to solve too many problems simultaneously.  Specifically, there are (at least) two issues to resolve:

1. initialization and invalid backpointers
2. handling nans/infs in the cost matrix

The two are related, as bad handling of nans/infs can lead the backtracker into states that expose invalid backpointers, which causes segfaults and initiated this thread in the first place.

The initialization behavior in the current master branch (fill with -1) is most certainly wrong, and is a prime example of using a sentinel value that's easily mistaken for real data: negative values are valid indices in numpy arrays, but we were using it as if it was not.

The fix as implemented so far in this PR ensures that backpointers are always valid (with respect to the steps matrix), even when a user supplies their own steps matrix.  In the latter case, the weights on the default steps are modified such that costs become infinite, implying that these steps are never selected in the optimization.  Of course, these infinite-cost steps are still "invalid" from the perspective of DTW since they use steps other than those defined by the user; all we're doing here is buying some safety from the backtracking routine to avoid segmentation faults in case it lands up in one of these situations, eg with a cost matrix (`C`) like:

```
0   inf 0    0
0   0    inf 0
0   0    inf 0
```
Here, no finite, full-sequence warping path exists, but it's not trivial to detect this situation until you get to the backtracking phase.

The accumulated cost should propagate these infs forward so they become easily detectable by checking the end-point of the warping path, as @fzalkow has implemented in #1033.  I think both changes are necessary to fully resolve the issue in this PR.